### ISSUE TITLE
WO-008: observation

### ISSUE THREAD

**Issue by Xocas12** [Xocas12/gosplan-env#13]
> **WO-008** - Phase 1 - Owner **MID-fast** - Difficulty **2/5** - Gate **G0**
> Card: [`workorders/WO-008.md`](https://github.com/Xocas12/gosplan-env/blob/main/workorders/WO-008.md)

## What this is

This card builds the vector an enterprise actually sees: `12 + 3J` components covering its own
target, stock, capital, last report, last audit outcome, last penalty, delivery fill and per-good
input coverage. Three bodies in `gosplan/env/obs.py` - the ordered component names, the `(N, d)`
observation array, and a phase mask.

It exists to enforce CONTRACT rule 6 at the agent boundary. `NEVER_OBSERVED` lists, as data, every
quantity that may not appear in any observation in any phase, and the behavioural test iterates it.
Two entries are subtler than they look: the current period's audit selection is forbidden, which is
what makes the audit an actual gamble; and periods remaining is forbidden because the horizon is
geometric, so any field correlating with it smuggles an end-game back in.

## Depends on

- WO-003
- WO-004
- WO-007

WO-003 is the dependency PLAN section 12.3 records. In practice also WO-004 (the `selfobs` draw) and
WO-007 (`reward_scale`, which scales observation field 9), which is why the DAG of PLAN section 12.2
puts WO-005 to WO-008 in one fan ahead of WO-009. The frozen suite is WO-002's.

**Blocked by:** #8 (WO-003), #9 (WO-004), #12 (WO-007)
This card may not be started until every issue above is closed (CONTRACT rule 12 / the loop in CONTRIBUTING.md).

## Read first (exhaustive whitelist)

- [`spec/spec.py`](https://github.com/Xocas12/gosplan-env/blob/main/spec/spec.py) - the frozen
  `obs_spec` signature and the PLAN section 2.4 layout in its docstring.
- [`gosplan/env/obs.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/env/obs.py) - **the
  write target.** `OBS_TABLE`, `SCALAR_FIELDS`, `PER_GOOD_BLOCKS`, `N_SCALAR_FIELDS`,
  `N_PER_GOOD_BLOCKS`, `NEVER_OBSERVED` and the full contract of `obs_spec`, `build_observation` and
  `phase_mask` are already there; three bodies missing.
- [`gosplan/config.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/config.py) -
  `EnvConfig`, the `Phase` literal, and the fields this card reads: `n_sectors`, `sector_of`,
  `productivity`, `steps_per_period`, `growth_directive`, `initial_target_frac`, `self_obs_noise`,
  `horizontal_visibility`.
- [`gosplan/rng.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/rng.py) - `draw` and
  `PURPOSES`; purpose `selfobs` is this card's.
- [`gosplan/env/state.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/env/state.py) -
  `State` fields and `INITIAL_CAPACITY` (`Kap_0`).
- [`gosplan/env/reward.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/env/reward.py) -
  `reward_scale`, which scales observation field 9.
- [`CONTRACT.md`](https://github.com/Xocas12/gosplan-env/blob/main/CONTRACT.md) - rule 6 (WELFARE
  BLINDNESS) above all, plus rules 3, 9, 12.
- [`pyproject.toml`](https://github.com/Xocas12/gosplan-env/blob/main/pyproject.toml) - ruff
  line-length 100, target py312.
- [`tests/unit/test_obs.py`](https://github.com/Xocas12/gosplan-env/blob/main/tests/unit/test_obs.py) -
  read-only (CONTRACT rule 2), from WO-002.
- [`tests/behavioural/test_welfare_blindness.py`](https://github.com/Xocas12/gosplan-env/blob/main/tests/behavioural/test_welfare_blindness.py) -
  read-only (CONTRACT rule 2), from WO-002.

Read nothing else. CONTRACT rule 12.

## Write only

- `gosplan/env/obs.py`

Creating, editing or deleting any other file is a CONTRACT rule 12 violation.

## Must pass

- [ ] `tests/unit/test_obs.py` - layout equals `obs_spec(cfg)`; dimension is `12 + 3J` at Phase 1;
      masks by phase; `need = 0` gives a coverage field of `1.0`; the `self_obs_noise = 0` case is
      exact.
- [ ] `tests/behavioural/test_welfare_blindness.py` - **T-B5**: sentinel `welfare`, sentinel
      `val_measured`, other enterprises' `y`, the current period's audit selection and periods
      remaining appear in NO observation.

```bash
uv run pytest tests/unit/test_obs.py -q
uv run pytest tests/behavioural/test_welfare_blindness.py -q
uv run ruff check gosplan/env/obs.py
uv run ruff format --check gosplan/env/obs.py
```

## Implementation notes

Three bodies. The layout data (`OBS_TABLE`, `SCALAR_FIELDS`, `PER_GOOD_BLOCKS`, `NEVER_OBSERVED`) is
already in the module and **is** the interface: `obs_spec` composes its answer from it, the frozen
test compares against it, and the PPO adapter (WO-017) slices observations by it. Changing a name, a
width or an order changes the agent interface and, after the v1 freeze, needs a `spec/CHANGELOG.md`
entry (CONTRACT rule 1).

### 1. `obs_spec(cfg) -> list[str]`

The ordered component names, length `12 + 3 * J` in Phase 1 with `J = cfg.supply.n_sectors`,
**composed from the module's table - not retyped as a literal list**:

```
0            phase                     0 produce / 1 report
1            k_over_M                  step within the period
2            log_target_ratio          log(T_i / T_0)
3            growth_directive          g; constant per config
4            cum_output_over_target    own true production so far this period
5            stock_over_target         S_i / T_i
6            capital_ratio             Kap_i / Kap_0
7            last_report_ratio
8            last_audited
9            last_penalty_scaled       last_penalty * reward_scale(cfg)
10           last_fill
11           inputs_delivered_total    delivered this period / need, totals
12   : 12+J  input_cov_{j}             X_ij / need_ij, per good
12+J : 12+2J sector_onehot_{j}         sector identity, for parameter sharing
12+2J: 12+3J deliv_cov_{j}             deliv_ij / need_ij this period, per good
(Phase 2, gated by `horizontal_visibility`) other enterprises' last report ratios in own sector
```

### 2. `build_observation(state, cfg, deliv, need) -> Array`

Returns `(N, d)` with `d = len(obs_spec(cfg))`, in exactly that order. `deliv` `(N, J)` is the
physical receipt from `gosplan.env.planner.deliver` and `need` `(N, J)` is the period's planned input
need `need_ij = a_{s(i)j} * T_i` - the same quantity the allocation weights of PLAN section 2.7.2 key
on. **Both are passed in** by the step machine (WO-009) and **must not be recomputed here**, so the
observation can never disagree with the delivery that produced it.

Per enterprise `i`, with `M = cfg.incentive.steps_per_period`,
`T_0 = cfg.tech.initial_target_frac * A_{s(i)} * cap_i` and `Kap_0 = INITIAL_CAPACITY`:

```
0  0.0 at a PRODUCE step, 1.0 at the REPORT step
1  state.k_step / M
2  log(state.target[i] / T_0[i])
3  cfg.incentive.growth_directive, the same value for every i
4  state.cum_output[i] / state.target[i]
5  state.inv_output[i] / state.target[i]
6  state.capital[i] / Kap_0[i]
7  state.last_report_ratio[i]
8  state.last_audited[i] as 0.0 / 1.0
9  state.last_penalty[i] * reward_scale(cfg)
10 state.last_fill[i]
11 sum_j deliv[i, j] / sum_j need[i, j]
12   : 12+J  state.inv_inputs[i, j] / need[i, j]
12+J : 12+2J 1.0 at j = cfg.supply.sector_of[i], else 0.0
12+2J: 12+3J deliv[i, j] / need[i, j]
```

**`need = 0` gives a coverage field of 1.0.** Wherever `need[i, j] == 0` - a good the enterprise's row
of `a` does not call for - `input_cov_j` and `deliv_cov_j` are `1.0`: not a division by zero, not a
NaN, not `0.0`. A good that is not needed is fully covered by definition, exactly as `coverage` in
`gosplan/env/production.py` treats it. The same rule applies to field 11 when `sum_j need[i, j] == 0`.

**Self-observation noise.** When `cfg.information.self_obs_noise = sigma > 0`, **fields 4 and 5
only** are multiplied by `exp(N(0, sigma**2))`, drawn through `gosplan.rng.draw` with purpose
`selfobs` and key `(seed_env, "selfobs", t, k, i)` (CONTRACT rule 9). No other field is noised. Phase
1 sets `sigma = 0.0`, which makes those two fields exact - the agent observes `S_i` and `y_i` exactly
at the REPORT step, so the report of PLAN section 2.8 is a choice made under full knowledge of the
truth. The branch must exist anyway: `self_obs_noise` is a mechanism toggle of the shape study over
`[0, 0.05]`.

### 3. `phase_mask(cfg, phase) -> Array`

A float mask `(d,)` carrying `1.0` for every component the period schedule of PLAN section 2.5 has
already defined at that phase and `0.0` otherwise. It is a pure function of `(cfg, phase)` and
**never touches a `State`**; `build_observation` applies it as its final operation, so a masked
component is `0.0` rather than stale.

Stated so nobody goes looking for a rule that is not there: PLAN section 2.4 tabulates **no per-field
phase gate**, and by the schedule of PLAN section 2.5 every one of the `12 + 3J` Phase-1 components
is already defined at both phases - DELIVER runs at the head of the period and fixes `last_fill`,
`deliv_cov_j` and `inputs_delivered_total`; the `last_*` fields are the previous period's outcome;
`cum_output` is a partial sum well defined from step 0. So **the Phase-1 mask is all ones at both
phases**, and the phase itself is carried by components 0 and 1 as values rather than by the mask.

The helper exists so the Phase-2 blocks that genuinely are gated - the peer block of `OBS_TABLE`,
gated by `information.horizontal_visibility` - have one place to be switched off. If
`tests/unit/test_obs.py` expects a different convention (for example a mask over ACTION dimensions
rather than observation components), **file an AMBIGUITY REPORT and stop**: do not special-case the
implementation, and do not invent a per-field gate.

A masked component is one the agent must not condition on. **Masking is not an information
mechanism**: the information arms of PLAN section 3 are `report_lag`, `aggregation_level`,
`channel_noise`, `self_obs_noise` and `horizontal_visibility`, and each is applied where it is
defined, not here.

### Rule 6 at the agent boundary

`NEVER_OBSERVED` lists, as data, every quantity that may not appear in any observation in any phase,
and T-B5 iterates it. Two entries are subtler than they look: the **audit selection for the current
period** is forbidden (an agent learns it was audited only through `last_audited` and
`last_penalty_scaled`, last period's outcome - which is what makes the audit an actual gamble); and
**periods remaining** is forbidden because the horizon is geometric (finding F4), so any field
correlating with it smuggles an end-game back in, which T-B9 regresses for.

## Forbidden

- Adding, removing, renaming or reordering any observation component, or changing a block width.
- Writing anything from `NEVER_OBSERVED` into the observation **by any route**, including derived
  ones: no function of `consumer_delivery` or welfare, no function of another enterprise's
  `cum_output`, `inv_output` or `inv_inputs`, no field carrying the current period's audit selection,
  no field correlating with periods remaining.
- Reading another enterprise's row. `build_observation` reads the true state strictly on the acting
  enterprise's own behalf, row by row; it is, with `make_planner_view`, one of only two sanctioned
  readers of `State`.
- Recomputing `deliv` or `need` inside this module instead of using the arguments.
- Any direct `numpy.random` or `jax.random` call (CONTRACT rule 9); the `selfobs` draw goes through
  `gosplan.rng.draw`.
- Noising any field other than 4 and 5, or applying the noise when `self_obs_noise == 0` in a way
  that changes the value.
- Inventing a per-field phase gate, or making `phase_mask` depend on a `State`.
- Importing scipy, pandas, jax or torch; importing `gosplan.env.step`, `gosplan.env.env` or any agent
  module (they import this one).
- Changing any signature, module constant, or module docstring; editing any test, `spec/spec.py`, or
  any module not on the "Write only" list.

## Contract rules that bite here

- **Rule 3 - STOP AND REPORT.** Choosing "the reasonable default" is a violation: do not add a field
  that would obviously help the agent, and do not invent a per-field phase gate PLAN section 2.4 does
  not state. If the frozen test expects another mask convention, stop and file.
- **Rule 12 - WORK ORDERS (whitelist).** Read only the ten entries above, write only
  `gosplan/env/obs.py`, run the completion command verbatim.
- **Rule 6 - WELFARE BLINDNESS.** This module is where the rule is enforced at the agent boundary:
  `welfare_true` and `val_measured` never appear in an observation, and neither does anything in
  `NEVER_OBSERVED`, by any derived route.
- **Rule 9 - RNG.** The `selfobs` draw goes through `gosplan.rng.draw`; no `numpy.random`, no
  module-level generator.
- **Rule 1 - FROZEN SPEC.** Component names, widths and order are the agent interface; changing one
  after the v1 freeze needs a `spec/CHANGELOG.md` entry.
- **Rule 2 - FROZEN TESTS.** Both test files are read-only; a disagreement is an AMBIGUITY REPORT.

## Definition of done

- [ ] `tests/unit/test_obs.py` green.
- [ ] `tests/behavioural/test_welfare_blindness.py` (T-B5) green.
- [ ] `uv run ruff check . && uv run ruff format --check .` clean.
- [ ] `uv run python scripts/contract_guard.py --all` clean.
- [ ] No file outside "Write only" touched.
- [ ] CI `ci-ok` green on the PR.
- [ ] The four-part report posted as the PR body: 1) tests passed/failed with output; 2) ambiguity
      reports filed, each listed, or "none"; 3) files changed - expect exactly `gosplan/env/obs.py`;
      4) anything you were tempted to invent and did not - in particular any phase gate, any extra
      field that would have helped the agent, and any handling of `need = 0` other than `1.0`.

**Comment by Xocas12** [comment]
Blocked on #53 — `obs.py` specifies the `selfobs` key as `(seed_env, "selfobs", t, k, i)` while `rng.py` states the trailing index is vectorised through `shape`. Both files are on this card's whitelist and they imply different random streams. Invisible in Phase 1 (`self_obs_noise = 0.0`).
